mainnet_contract/contracts/Subnet.sol

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.4.21 <0.9.0;
+pragma experimental ABIEncoderV2;
+
+import "./HeaderReader.sol";
+
+contract Subnet {
+
+  struct Header {
+    bytes32 hash;
+    int number;
+    uint64 round_num;
+    bytes32 parent_hash;
+    bool finalized;
+    uint mainnet_num;
+    bytes src;
+  }
+
+  struct Validators {
+    address[] set;
+    int threshold;
+  }
+
+  mapping(bytes32 => Header) header_tree;
+  mapping(int => Validators) validator_sets;
+  mapping(address => bool) lookup;
+  mapping(address => bool) unique_addr;
+  mapping(address => bool) masters;
+  int current_validator_set_pointer = 0;
+  int current_subnet_height;
+  bytes32 latest_finalized_block;
+
+  // Event types
+  event SubnetBlockAccepted(bytes32 block_hash, int number);
+  event SubnetBlockFinalized(bytes32 block_hash, int number);
+
+  // Modifier
+  modifier onlyMasters() {
+    if (!masters[msg.sender]) revert("Masters Only");
+    _;
+  }
+
+  constructor(address[] memory initial_validator_set, int threshold, bytes memory genesis_header) public {
+    require(initial_validator_set.length > 0, "Validator Empty");
+    bytes32 genesis_header_hash = keccak256(genesis_header);
+    (bytes32 ph, int n) = HeaderReader.getParentHashAndNumber(genesis_header);
+    header_tree[genesis_header_hash] = Header({
+      hash: genesis_header_hash,
+      number: n,
+      round_num: 0, 
+      parent_hash: ph,
+      finalized: true,
+      mainnet_num: block.number,
+      src: genesis_header
+    });
+    validator_sets[0] = Validators({
+      set: initial_validator_set,
+      threshold: threshold
+    });
+    for (uint i = 0; i < initial_validator_set.length; i++) {
+      lookup[initial_validator_set[i]] = true;
+    }
+    masters[msg.sender] = true;
+    latest_finalized_block = genesis_header_hash;
+  }
+
+  function isMaster(address master) public view returns (bool) {
+    return masters[master];
+  }
+
+  function addMaster(address master) public onlyMasters {
+    masters[master] = true;
+  }
+
+  function removeMaster(address master) public onlyMasters {
+    masters[master] = false;
+  }
+
+  function reviseValidatorSet(address[] memory new_validator_set, int threshold, int subnet_block_height) public onlyMasters  {
+    require(new_validator_set.length > 0, "Validator Empty");
+    require(threshold > 0, "Invalid Threshold");
+    require(subnet_block_height >= current_validator_set_pointer, "Modify Past Validator");
+    validator_sets[subnet_block_height] = Validators({
+      set: new_validator_set,
+      threshold: threshold
+    });
+  }
+
+  function receiveHeader(bytes memory header) public onlyMasters { 
+    (bytes32 parent_hash, int number, uint64 round_number, bytes32 signHash, bytes[] memory sigs) = HeaderReader.getValidationParams(header);
```

Comment:
read two more: currentEpochValidators, next...

- currentEpochValidators = nil for all !IsEpochSwitch(block).
- nextEpochValidators = nil for all block: number % EPOCH != EPOCH-GAP(+1).

Need to check: when encounter currentEpochValidators, check == last committed nextEpochValidators